lib/rubygems/exceptions.rb

```diff
@@ -20,6 +20,53 @@ class Gem::Exception < RuntimeError
 
 class Gem::CommandLineError < Gem::Exception; end
 
+class Gem::UnknownCommandError < Gem::Exception
+
+  class SpellChecker
+
+    attr_reader :error
+
+    def initialize(error)
+      @error = error
+    end
+
+    def corrections
+      @corrections ||=
+        spell_checker.correct(error.command_name).map(&:inspect)
+    end
+
+    private
+
+    def spell_checker
+      DidYouMean::SpellChecker.new(dictionary: error.command_names)
+    end
+
+  end
+
+  attr_reader :command_names, :command_name
+
+  def initialize(command_names, command_name)
```

Comment:
I think `command_name` should be more descriptive like `unknown_name`.

I don't think you need to pass in `command_names` at all because CommandManager is a singleton and you can `Gem::CommandManager.instance.command_names` when you need to extract the dictionary in SpellChecker above.  Holding the data in this class doesn't seem useful.

Reply:
Okay, that's fine I've made the update.